### ISSUE TITLE
No operator highlighting

### ISSUE THREAD

**Issue by quakerquickoats** [ocaml/tuareg#207]
Previously I had tuareg from melpa-stable. I've upgraded all my packages to melpa (unstable), tried tuareg 20190513.2105 from there, and the font-lock on operators has disappeared. I've explored all of the customize options and even some other elisp code from the web for generic font lock on operators. No luck. I've deleted the package, and instead installed tuareg using opam and the font locking of operators has returned.

[This is a very important feature for me, and I am not sure why it is rare to see any emacs modes support this.]

**Comment by rgrinberg** [comment]
@Chris00 has done some extremely important work in making tuareg's font locking performant on large buffers and various corner cases. So I just want to say that if there's some loss of highlighting as a result of the optimization, then I'm all in favor of this trade-off.

**Comment by quakerquickoats** [comment]
I understand that you are in favor of the inverse of my issue, and I appreciate that you are somewhat confident in your personal requirements in the disappearance of this feature. Your reply is discouraging and slightly disheartening, and unexpected. I suggest that for the benefit of others like me to let them know about this, rather than the surprise of perceivably downgrading situation for most users. I wouldn't want users to spend an hour of their own time figuring out if they've done something wrong or if something else was broken such as Emacs itself. I do not understand how this could not be an option - if someone were to be working on large buffers where font-lock is not fast enough, which is not exactly a common situation. I would like it if you could explain "corner cases" if tuareg-mode 2.2 works just fine. I will have a look at a diff afterwards and see if there is a simple solution for everyone.

I will stick with 2.2 tuareg from opam for the foreseeable future.

**Comment by rgrinberg** [comment]
Sure, I think it's reasonable to have extra highlighting turned for those who aren't worried about performance. However, for the default experience it's quite frustrating to have tuareg get stuck for seconds on every keystroke in some large buffers.

In the other thread, you've mentioned that such large buffers aren't reasonable and that you would split them into multiple files if you could. Well in fact, this isn't always reasonable. First of all, it's quite common that one is browsing source that is outside of your control. Second of all, splitting a compilation unit in OCaml can be far more trivial. It has implications such as introducing circular dependencies and introducing implementation details in the interface where non have existed before. You can check out many mature open source packages and easily convince yourself that working with large buffers is commonplace.

**Comment by jberdine** [comment]
This change was the topic of https://github.com/ocaml/tuareg/issues/204 . Things improved significantly, but I'm still in a don't upgrade situation. But the current highlighting of patterns is notably better than 2.2, so some hybrid would be attractive.

It would be interesting to know very roughly what values of "large" are that trigger the trouble.

**Comment by quakerquickoats** [comment]
I wonder if it is more of a problem with emacs' own internals when it comes to large buffers. I notice that large org-mode buffers are more responsive when completely folded in. I wonder if there is a way to only font lock certain areas, such as around cursor ? I understand of course there are scoping issues. But granularity could perhaps be a bit more dynamic like you've mentioned with how "large" is a large buffer. It really depends. Maybe timing inside of emacs for font-lock operations could reduce quality. I learned today that there was a recent addition of a 'font-lock-maximum-size' and 'font-lock-maximum-decoration' options. For example, the default is "maximum" for the latter, which could possibly enable operator-granular coloring. I think there could be then an easy way to have simply "font lock ops taking too long, reduce decorations? y/n/auto", where 'auto' here could be adjusting automatically according to buffer size as well. Just thoughts. 

Now, how come merlin is so fast? And too, with Eldoc mode enabled? I am interested what makes coloring so much work that could not be improved with emacs itself for all modes.

[Side remark, I've been thinking recently, what it would be like in general for coloring of any code, "context-sensitive" highlighting, as opposed to whole-file. Whereas especially with namespaces and scope that there is no global context aside from plenty of regexp.]

**Comment by jberdine** [comment]
And let me just say that I really appreciate the work improving efficiency. There is no question there. I just haven't been convinced that we can't have everything. But I haven't had time to work on it myself either.

**Comment by rgrinberg** [comment]
> It would be interesting to know very roughly what values of "large" are that trigger the trouble.

Large is perhaps not the right criteria. There are some large buffers that tuareg has no problems with, and some relatively modest buffers (~1k lines) where tuareg shows noticeable lag.

**Comment by rgrinberg** [comment]
> And let me just say that I really appreciate the work improving efficiency. There is no question there. I just haven't been convinced that we can't have everything. But I haven't had time to work on it myself either.

Indeed, but I think the right approach is gain back efficiency and then slowly start re-introducing highlighting. The marginal benefit of a few extra colors is quite low to being unable to efficiently edit some buffers.

**Comment by Chris00** [comment]
As @rgrinberg said, let us finish the work on efficiency and then tackle this again.  It is not difficult to highlight _all_ operators (as opposed to the non-standard ones), the question is whether this should be the default or be optional.

**Comment by jberdine** [comment]
Ack, thanks. Also, I wonder what you think of trying the flipped situation: recognize and highlight the "identifiers" (i.e. 'ident' in [section 7.1](http://caml.inria.fr/pub/docs/manual-ocaml/lex.html) minus keywords, etc.) and leave everything else as "operators". Perhaps expressing it that way around can lead to faster regexps.

**Comment by Chris00** [comment]
I think it is desirable to distinguish function names form variables and module names from constructors.  Not doing so may indeed lead to faster regexps but IMHO syntax highlighting is useful to distinguish at a glance what can't be from their spelling so I am not too keen going into that direction.
The point of waiting that the rest is optimized is to be able to judge the cost of this feature (that is important to some developers but not all).  Its cost will determine whether it will be turned on by default.

**Comment by Chris00** [comment]
1. Full operator highlighting will be (per-buffer) optional as it is slow.
2. How much important is it that `=` be highlighted?  Currently, I already remove its fontification for `let ... =` (and `method ... =`) and in optional arguments of `let` but there is also `type ... =` and records `{ label = v ... }` (and maybe other places) where `=` is not an infix operator.

**Comment by Chris00** [comment]
Most common cases are now covered when `tuareg-highlight-all-operators` is `t`.  May you please try it and report?

**Comment by quakerquickoats** [comment]
No results, checked out latest branch; with (setq tuareg-highlight-all-operators t) , removed all other tuareg etc. Can send screens

**Comment by quakerquickoats** [comment]
Is it 'tuareg-font-lock-symbols' as per M-x customize ? Because this was also set to `t'.

**Comment by Chris00** [comment]
> `tuareg-font-lock-symbols`

This is to prettify operators and is orthogonal — even though it is handled by font-lock.

**Comment by Chris00** [comment]
> No results, checked out latest branch; with (setq tuareg-highlight-all-operators t) , removed all other tuareg etc. 

Did you (re)compile and point Emacs to your repo?  If you see nothing with [`tuareg-highlight-all-operators`](https://github.com/ocaml/tuareg/blob/master/tuareg.el#L1129), it is because old code is being used.  You can see which one by evaluating (<kbd>C-xC-e</kbd>) the phase `(locate-library "tuareg")` (copy it to `*scratch*` first).

**Comment by quakerquickoats** [comment]
Yep, definately double-checked tuareg.el was in the right place, restarted clean etc. (Had to do this before to make sure I regained operator highlights from opam when i first [up]graded to melpa)

I do C-h f <RET> tuareg-mode and points me to the .el where it was require'd from. I had completely removed all previous tuareg before trying. Won't mind trying again soon and will report results.

**Comment by Chris00** [comment]
Did you set `tuareg-highlight-all-operators` in the mode hook?  If so, set it outside, the hook is executed last (as usual)—after font lock is set.

**Comment by quakerquickoats** [comment]
Could you elaborate on this please ? I've no mode hooks set for tuareg. =) aside from merlin.

**Comment by Chris00** [comment]
I mean you have 
```elisp
(setq tuareg-highlight-all-operators t)
```
in your [Init file](https://www.gnu.org/software/emacs/manual/html_node/emacs/Init-File.html)—at the “toplevel”.  (If set in the Tuareg hook, it will be executed after font-lock is initialized and will therefore have no effect on the first opened file.)

**Comment by Chris00** [comment]
After loading an `.ml` file, you can also evaluate
```elisp
(setq tuareg-highlight-all-operators t)
```
and then restart the mode with <kbd>M-x tuareg-mode</kbd>.  If the right elisp code is loaded, the highlighting of operators should happen.

**Comment by quakerquickoats** [comment]
Sorry, I have been away from my project. Yes this is what I have done. I've tried again just now, using latest sources from git here, it is not working still yet. Is there something I am missing? I've got emacs 27

**Comment by Chris00** [comment]
This is very strange.  I'm using Emacs 26.  I'll test with 27.

**Comment by Chris00** [comment]
@quakerquickoats Did you find the culprit on your machine?  Here is a screenshot with prettify-symbols and `(setq tuareg-highlight-all-operators t)`.  (As you see, both work well together.)

![Operators](https://user-images.githubusercontent.com/1255665/70394340-a63dbd00-19f4-11ea-908c-d1df54ef4a91.png)
